File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        r = len(candidates) - 1
        while r >= 0 and candidates[r] > target:
            r -= 1
        if r < 0:
            return res

        def pathSum(left: int, path: List[int], sumUntil: int) -> (bool, List[int]):
            if sumUntil == target:
                res.append(path)
                return (True, path)
            elif sumUntil > target:
                return (False, path)
            while left <= r and target >= sumUntil + candidates[left]:
                pathSum(left, path + [candidates[left]], sumUntil + candidates[left])
                left += 1
            return (False, path)

        for l in range(r+1):
            pathSum(l, [candidates[l]], candidates[l])
        return res

File: 39-combination-sum/combination-sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const result: number[][] = [];
    
    // Helper function for backtracking
    function backtrack(start: number, target: number, current: number[]) {
        // Base case: if target becomes 0, we found a valid combination
        if (target === 0) {
            result.push([...current]);
            return;
        }
        
        // If target becomes negative, no need to continue this path
        if (target < 0) {
            return;
        }
        
        // Try each candidate from start index
        for (let i = start; i < candidates.length; i++) {
            // Add current candidate to combination
            current.push(candidates[i]);
            // Recursively try with same index (since we can reuse numbers)
            backtrack(i, target - candidates[i], current);
            // Backtrack by removing the last added number
            current.pop();
        }
    }
    
    // Start backtracking with empty combination
    backtrack(0, target, []);
    return result;
};